File: dev/requirements/logic/cases/segment-mode-availability.5.2.case.swift
import Foundation
import SetlistCore

/// 5.2 — The segment-mode switch cannot be turned on while no autodetector is
/// available, and says so rather than silently ignoring the gesture.
let segmentModeAvailability_5_2 = RequirementCase(id: "5.2") {
    let jokes = [Joke(title: "Gym induction", body: "…")]

    var phaseA = StageModel(jokes: jokes, plannedLength: 1200)
    try expectEqual(phaseA.segmentMode, .manual, "the switch starts off")
    do {
        try phaseA.setSegmentMode(.autodetect)
        throw RequirementFailure(description: "autodetect was accepted with no detector registered")
    } catch let refusal as SegmentModeRefusal {
        try expectEqual(refusal, .autodetectUnavailable, "refusal reason")
    }
    try expectEqual(phaseA.segmentMode, .manual, "the switch stayed off")

    // The view needs no change once a detector registers.
    var withDetector = StageModel(jokes: jokes, plannedLength: 1200, autodetectAvailable: true)
    try withDetector.setSegmentMode(.autodetect)
    try expectEqual(withDetector.segmentMode, .autodetect, "the switch turns on once a detector exists")
}

File: dev/requirements/logic/cases/segment-mode-availability.5.2.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 5.2 — The switch cannot be turned on while no autodetector is available,
/// and reports itself unavailable rather than silently ignoring the gesture.
final segmentModeAvailability_5_2 = RequirementCase('5.2', (tester) {
  final without = StageModel(jokes: const [], plannedLength: const Duration(minutes: 20));

  var refused = false;
  try {
    without.setSegmentMode(SegmentMode.autodetect);
  } on SegmentModeRefusal {
    refused = true;
  }

  expectTrue(refused, 'turning it on is refused, not ignored');
  expectThat(
    without.segmentMode,
    SegmentMode.manual,
    'and the mode does not half-change behind the refusal',
  );

  final available = StageModel(
    jokes: const [],
    plannedLength: const Duration(minutes: 20),
    autodetectAvailable: true,
  );
  available.setSegmentMode(SegmentMode.autodetect);
  expectThat(available.segmentMode, SegmentMode.autodetect, 'a registered detector makes it real');
});
